truffleruby/src/main/ruby/core/truffle/string_operations.rb

```diff
@@ -0,0 +1,110 @@
+module Truffle
+  class StringOperations
+
+    def self.gsub_block_set_last_match(s, pattern, &block)
+      Truffle::StringOperations.gsub_internal(s, pattern) do |m, str|
+        Regexp.set_block_last_match(block, m)
+        yield str
+      end
+    end
+
+    def self.gsub_internal(orig, pattern, replacement=undefined, &block)
+      unless orig.valid_encoding?
+        raise ArgumentError, "invalid byte sequence in #{orig.encoding}"
+      end
+
+      if undefined.equal? replacement
+        use_yield = true
```

Comment:
I've gone for a more complete refactor since use_yield wan't the only thing that really needed cleaning up in that code.